### DISCUSSION TITLE
setTimeout for 30s, not working on vercel , but works locally

### DISCUSSION THREAD

**Discussion by skushagra9** [vercel/next.js#66315] [Help]
### Summary

I have a setTimeout , on my serverless api for 30 mins.
When I run it locally , it works, but when I run it hosted on vercel, this function doesn't runs. Even It doesn't gives any error.

I tried to do it - by setting the timeout for 2 mins, and it worked, but why not for 30 mins ???

**Comment by icyJoseph** [comment]
- https://vercel.com/docs/functions/runtimes#max-duration

**Reply by skushagra9** [reply]
> * https://vercel.com/docs/functions/runtimes#max-duration

but i dont think this is the problem, it doesn't gives out any error- but the function doesn't executes after specific timeout

**Reply by icyJoseph** [reply]
But you return from the handler, and let the timer callback running?

**Reply by skushagra9** [reply]
yes

**Reply by skushagra9** [reply]
it works fine locally, but why vercel gives the problem ??

**Reply by icyJoseph** [reply]
Yeah, so since you already returned, unless more requests come through, and sometimes, even in spite of that, the task will be killed, after some time.

**Reply by icyJoseph** [reply]
For example, and this won't help you I think, but in middleware they have this helper:

- https://nextjs.org/docs/app/building-your-application/routing/middleware#waituntil-and-nextfetchevent

That would hint at not killing the task until that work resolved.

This also won't really help you, https://nextjs.org/blog/next-15-rc#executing-code-after-a-response-with-nextafter-experimental, but that's an experimental way they are preparing, to allow the framework to do work, after, and without blocking sending a response to the user.

Depending on what you want to do, maybe a cronjob is your call here, or maybe consider using other service providers. This can be done in different levels, like keep everything else in Vercel, but host elsewhere a server that you'd call from your Vercel deployed app, and that'd schedule a 30 minutes task, or move everything to those providers. 

- https://www.shuttle.rs/pricing
- https://render.com/
- https://railway.app/

Etc... keep in mind that, having a server instance running, or just idle for long periods of time, is gonna have to be a cost for someone, so finding a free alternative, is not gonna be trivial, though there are some out there.

**Reply by skushagra9** [reply]
Hi @icyJoseph , 
Thanks a lot for you support , I am implementing a cron job for this now.
I feel people like you are really important for the community and keeps everyone going.

**Comment by skushagra9** [comment]
got it , thats why it works when I set it for 2 mins right
what would be the best approach here ??

**Reply by skushagra9** [reply]
@icyJoseph
